src/pykeen/triples/triples_factory.py

```diff
@@ -163,6 +169,291 @@ def _map_triples_elements_to_ids(
     return torch.tensor(unique_mapped_triples, dtype=torch.long)
 
 
+def _check_already_inverted_relations(relations: Iterable[str]) -> bool:
+    return any(relation.endswith(INVERSE_SUFFIX) for relation in relations)
+
+
+def _map_column(
+    labeled_column: Union[np.ndarray, Sequence[str], str],
+    mapper: Callable[[np.ndarray, Tuple[int]], np.ndarray],
+    unknown_id: int = -1,
+) -> torch.LongTensor:
+    """Apply a vectorized mapping to a column of labels.
+
+    :param labeled_column:
+        The column of labels.
+    :param mapper:
+        The vectorized label-to-id mapper.
+    :param unknown_id:
+        An ID to use for unknown labels.
+
+    :return:
+        An array of IDs.
+    """
+    # Input normalization
+    if isinstance(labeled_column, str):
+        labeled_column = [labeled_column]
+    labeled_column = np.asanyarray(labeled_column)
+    mapped_column = mapper(labeled_column, (unknown_id,))
+    return torch.as_tensor(data=mapped_column, dtype=torch.long)
+
+
+def _label_column(
+    mapped_column: Union[np.ndarray, Sequence[int], torch.LongTensor, int],
+    vectorized_labeler: Callable[[np.ndarray, Tuple[str]], np.ndarray],
+    unknown_label: str,
+) -> np.ndarray:
+    """Apply vectorized labeler to an array of IDs.
+
+    :param mapped_column:
+        The IDs.
+    :param vectorized_labeler:
+        The vectorized labeler.
+    :param unknown_label:
+        A label to use for unknown IDs.
+
+    :return:
+        An array of labels, same shape as mapped_column.
+    """
+    # Input normalization
+    if isinstance(mapped_column, int):
+        mapped_column = [mapped_column]
+    if torch.is_tensor(mapped_column):
+        mapped_column = mapped_column.cpu().numpy()
+    mapped_column = np.asanyarray(mapped_column)
+
+    # Actual labeling
+    return vectorized_labeler(mapped_column, (unknown_label,))
+
+
+@dataclass
+class LabelMapping:
+    """A mapping from labels to IDs."""
+
+    #: The mapping for entities
+    entity_label_to_id: EntityMapping
+
+    #: The mapping for relations
+    relation_label_to_id: RelationMapping
+
+    #: A dictionary mapping each relation to its inverse, if inverse triples were created
+    relation_to_inverse: Optional[Mapping[str, str]] = None
+
+    @staticmethod
+    def from_labeled_triples(triples: LabeledTriples) -> 'LabelMapping':
+        """Create a mapping from labeled triples."""
+        return LabelMapping(
+            entity_label_to_id=create_entity_mapping(triples=triples),
+            relation_label_to_id=create_relation_mapping(relations=triples[:, 1])
+        )
+
+    @property
+    def entity_id_to_label(self) -> Mapping[int, str]:
+        """The mapping from entity IDs to labels."""
+        return invert_mapping(mapping=self.entity_label_to_id)
```

Comment:
should these be cached? We can set up a `__post_init__` function to add these automatically

Reply:
or use `functools.lru_cache` to make them lazy

Reply:
I am not sure whether the computation time savings justify the additional memory overhead of storing redundant information.

Reply:
cf. dac6e5c